httpclient5/src/main/java/org/apache/hc/client5/http/impl/nio/HappyEyeballsV2AsyncClientConnectionOperator.java

```diff
@@ -0,0 +1,680 @@
+                    return CompletableFuture.anyOf(attempts.toArray(new CompletableFuture[0]))
+                            .thenCompose(result -> {
+                                if (result instanceof ManagedAsyncClientConnection) {
+                                    // If there is a result, cancel all other attempts and complete the connectionFuture
+                                    connectionFutures.forEach(future -> future.cancel(true));
+                                    connectionFuture.complete((ManagedAsyncClientConnection) result);
+                                } else {
+                                    // If there is an exception, complete the connectionFuture exceptionally with the exception
+                                    connectionFuture.completeExceptionally(new ConnectException("Failed to connect to any address for " + host));
+                                }
+                                // Invoke the callback if provided
+                                if (callback != null) {
+                                    connectionFuture.whenComplete((conn, ex) -> {
+                                        if (ex != null) {
+                                            callback.failed(new Exception(ex));
+                                        } else {
+                                            callback.completed(conn);
+                                        }
+                                    });
+                                }
+                                return connectionFuture;
+                            });
+                })
+                .exceptionally(e -> {
+                    connectionFuture.completeExceptionally(e);
+                    if (callback != null) {
+                        callback.failed(new Exception(e));
+                    }
+                    return null;
+                });
+
+        return connectionFuture;
+    }
+
+    /**
+     * Asynchronously resolves the DNS for the given host name and returns a CompletableFuture that will be completed
+     * with an array of InetAddress objects representing the IP addresses of the host.
+     * The resolution of AAAA records is delayed by the configured resolution delay to allow for a chance for A records to be
+     * returned first.
+     *
+     * @param host the host name to resolve DNS for
+     * @return a CompletableFuture that will be completed with an array of InetAddress objects representing the IP addresses
+     */
+    private CompletableFuture<InetAddress[]> resolveDnsAsync(final String host) {
+        final CompletableFuture<InetAddress[]> dnsFuture = new CompletableFuture<>();
+        CompletableFuture.runAsync(() -> {
+            try {
+                final InetAddress[] inetAddresses = dnsResolver.resolve(host);
+                // Introduce a delay before resolving AAAA records after receiving A records
+                resolution_delay.sleep();
```

Comment:
Instead of using Thread.sleep(), i could use non-blocking methods(ScheduledExecutorService) to delay the execution of the next connection attempt.

Reply:
Yes, in async code delays should always be implemented by yielding the current thread and scheduling a continuation, rather than calling `Thread.sleep`. Ideally, the DNS lookup would also be made as an async call, but to my knowledge there's really nothing we can do about that, and in practical terms DNS is very fast. I'm not even sure Netty has an async implementation of DNS lookups.

Reply:
Netty does, but it's an entirely bespoke implementation that doesn't touch the JDK's built-int resolver. The blocking here will hopefully become less of a problem for this client once Loom is available.